File: frontend/wms/lib/app/data/model/wms_caixa_model.dart
import 'dart:convert';
import 'package:pluto_grid/pluto_grid.dart';

import 'package:wms/app/data/model/model_imports.dart';

class WmsCaixaModel {
	int? id;
	int? idWmsEstante;
	String? codigo;
	int? altura;
	int? largura;
	int? profundidade;
	List<WmsArmazenamentoModel>? wmsArmazenamentoModelList;
	WmsEstanteModel? wmsEstanteModel;

	WmsCaixaModel({
		this.id,
		this.idWmsEstante,
		this.codigo,
		this.altura,
		this.largura,
		this.profundidade,
		this.wmsArmazenamentoModelList,
		this.wmsEstanteModel,
	});

	static List<String> dbColumns = <String>[
		'id',
		'codigo',
		'altura',
		'largura',
		'profundidade',
	];
	
	static List<String> aliasColumns = <String>[
		'Id',
		'Codigo',
		'Altura',
		'Largura',
		'Profundidade',
	];

	WmsCaixaModel.fromJson(Map<String, dynamic> jsonData) {
		id = jsonData['id'];
		idWmsEstante = jsonData['idWmsEstante'];
		codigo = jsonData['codigo'];
		altura = jsonData['altura'];
		largura = jsonData['largura'];
		profundidade = jsonData['profundidade'];
		wmsArmazenamentoModelList = (jsonData['wmsArmazenamentoModelList'] as Iterable?)?.map((m) => WmsArmazenamentoModel.fromJson(m)).toList() ?? [];
		wmsEstanteModel = jsonData['wmsEstanteModel'] == null ? WmsEstanteModel() : WmsEstanteModel.fromJson(jsonData['wmsEstanteModel']);
	}

	Map<String, dynamic> get toJson {
		Map<String, dynamic> jsonData = <String, dynamic>{};

		jsonData['id'] = id != 0 ? id : null;
		jsonData['idWmsEstante'] = idWmsEstante != 0 ? idWmsEstante : null;
		jsonData['codigo'] = codigo;
		jsonData['altura'] = altura;
		jsonData['largura'] = largura;
		jsonData['profundidade'] = profundidade;
		
		var wmsArmazenamentoModelLocalList = []; 
		for (WmsArmazenamentoModel object in wmsArmazenamentoModelList ?? []) { 
			wmsArmazenamentoModelLocalList.add(object.toJson); 
		}
		jsonData['wmsArmazenamentoModelList'] = wmsArmazenamentoModelLocalList;
		jsonData['wmsEstanteModel'] = wmsEstanteModel?.toJson;
	
		return jsonData;
	}
	
	String objectEncodeJson() {
		final jsonData = toJson;
		return json.encode(jsonData);
	}

	plutoRowToObject(PlutoRow plutoRow) {
		id = plutoRow.cells['id']?.value;
		idWmsEstante = plutoRow.cells['idWmsEstante']?.value;
		codigo = plutoRow.cells['codigo']?.value;
		altura = plutoRow.cells['altura']?.value;
		largura = plutoRow.cells['largura']?.value;
		profundidade = plutoRow.cells['profundidade']?.value;
		wmsArmazenamentoModelList = [];
		wmsEstanteModel = WmsEstanteModel();
		wmsEstanteModel?.codigo = plutoRow.cells['wmsEstanteModel']?.value;
	}	

	WmsCaixaModel clone() {
		return WmsCaixaModel(
			id: id,
			idWmsEstante: idWmsEstante,
			codigo: codigo,
			altura: altura,
			largura: largura,
			profundidade: profundidade,
			wmsArmazenamentoModelList: wmsArmazenamentoModelListClone(wmsArmazenamentoModelList!),
		);			
	}

	wmsArmazenamentoModelListClone(List<WmsArmazenamentoModel> wmsArmazenamentoModelList) { 
		List<WmsArmazenamentoModel> resultList = [];
		for (var wmsArmazenamentoModel in wmsArmazenamentoModelList) {
			resultList.add(
				WmsArmazenamentoModel(
					id: wmsArmazenamentoModel.id,
					idWmsCaixa: wmsArmazenamentoModel.idWmsCaixa,
					idWmsRecebimentoDetalhe: wmsArmazenamentoModel.idWmsRecebimentoDetalhe,
					quantidade: wmsArmazenamentoModel.quantidade,
				)
			);
		}
		return resultList;
	}

	
}

File: backend/wms/src/model/wms_caixa_model.py
from src import db
from src.model.wms_armazenamento_model import WmsArmazenamentoModel
from src.model.wms_estante_model import WmsEstanteModel


class WmsCaixaModel(db.Model):
    __tablename__ = 'wms_caixa'

    id = db.Column(db.Integer, primary_key=True)
    codigo = db.Column(db.String(10))
    altura = db.Column(db.Integer)
    largura = db.Column(db.Integer)
    profundidade = db.Column(db.Integer)
    id_wms_estante = db.Column(db.Integer, db.ForeignKey('wms_estante.id'))

    wms_armazenamento_model_list = db.relationship('WmsArmazenamentoModel', lazy='dynamic')
    wms_estante_model = db.relationship('WmsEstanteModel', foreign_keys=[id_wms_estante])

    def mapping(self, data):
        self.id = data.get('id')
        self.id_wms_estante = data.get('idWmsEstante')
        self.codigo = data.get('codigo')
        self.altura = data.get('altura')
        self.largura = data.get('largura')
        self.profundidade = data.get('profundidade')

    def serialize(self):
        return {
            'id': self.id,
            'idWmsEstante': self.id_wms_estante,
            'codigo': self.codigo,
            'altura': self.altura,
            'largura': self.largura,
            'profundidade': self.profundidade,
            'wmsArmazenamentoModelList': [wms_armazenamento_model.serialize() for wms_armazenamento_model in self.wms_armazenamento_model_list],
            'wmsEstanteModel': self.wms_estante_model.serialize() if self.wms_estante_model else None,
        }